File: csharp/Tennis/TennisTests.cs
using System;
using NUnit.Framework;

namespace Tennis
{
    [TestFixture( 0,  0, "Love-All")]
    [TestFixture( 1,  1, "Fifteen-All")]
    [TestFixture( 2,  2, "Thirty-All")]
    [TestFixture( 3,  3, "Deuce")]
    [TestFixture( 4,  4, "Deuce")]
    [TestFixture( 1,  0, "Fifteen-Love")]
    [TestFixture( 0,  1, "Love-Fifteen")]
    [TestFixture( 2,  0, "Thirty-Love")]
    [TestFixture( 0,  2, "Love-Thirty")]
    [TestFixture( 3,  0, "Forty-Love")]
    [TestFixture( 0,  3, "Love-Forty")]
    [TestFixture( 4,  0, "Win for player1")]
    [TestFixture( 0,  4, "Win for player2")]
    [TestFixture( 2,  1, "Thirty-Fifteen")]
    [TestFixture( 1,  2, "Fifteen-Thirty")]
    [TestFixture( 3,  1, "Forty-Fifteen")]
    [TestFixture( 1,  3, "Fifteen-Forty")]
    [TestFixture( 4,  1, "Win for player1")]
    [TestFixture( 1,  4, "Win for player2")]
    [TestFixture( 3,  2, "Forty-Thirty")]
    [TestFixture( 2,  3, "Thirty-Forty")]
    [TestFixture( 4,  2, "Win for player1")]
    [TestFixture( 2,  4, "Win for player2")]
    [TestFixture( 4,  3, "Advantage player1")]
    [TestFixture( 3,  4, "Advantage player2")]
    [TestFixture( 5,  4, "Advantage player1")]
    [TestFixture( 4,  5, "Advantage player2")]
    [TestFixture(15, 14, "Advantage player1")]
    [TestFixture(14, 15, "Advantage player2")]
    [TestFixture( 6,  4, "Win for player1")]
    [TestFixture( 4,  6, "Win for player2")]
    [TestFixture(16, 14, "Win for player1")]
    [TestFixture(14, 16, "Win for player2")]
    public class TennisTests
    {
        private readonly int player1Score;
        private readonly int player2Score;
        private readonly string expectedScore;

        public TennisTests(int player1Score, int player2Score, string expectedScore)
        {
            this.player1Score = player1Score;
            this.player2Score = player2Score;
            this.expectedScore = expectedScore;
        }

        [Test]
        public void CheckTennisGame1()
        {
            var game = new TennisGame1("player1", "player2");
            CheckAllScores(game);
        }

        [Test]
        public void CheckTennisGame2()
        {
            var game = new TennisGame2("player1", "player2");
            CheckAllScores(game);
        }

        [Test]
        public void CheckTennisGame3()
        {
            var game = new TennisGame3("player1", "player2");
            CheckAllScores(game);
        }

        private void CheckAllScores(ITennisGame game)
        {
            var highestScore = Math.Max(this.player1Score, this.player2Score);
            for (var i = 0; i < highestScore; i++)
            {
                if (i < this.player1Score)
                    game.WonPoint("player1");
                if (i < this.player2Score)
                    game.WonPoint("player2");
            }
            Assert.AreEqual(this.expectedScore, game.GetScore());
        }

    }

    [TestFixture]
    public class ExampleGameTennisTest
    {
        [Test]
        public void CheckGame1()
        {
            var game = new TennisGame1("player1", "player2");
            RealisticTennisGame(game);
        }

        [Test]
        public void CheckGame2()
        {
            var game = new TennisGame2("player1", "player2");
            RealisticTennisGame(game);
        }

        [Test]
        public void CheckGame3()
        {
            var game = new TennisGame3("player1", "player2");
            RealisticTennisGame(game);
        }

        private void RealisticTennisGame(ITennisGame game)
        {
            string[] points = { "player1", "player1", "player2", "player2", "player1", "player1" };
            string[] expectedScores = { "Fifteen-Love", "Thirty-Love", "Thirty-Fifteen", "Thirty-All", "Forty-Thirty", "Win for player1" };
            for (var i = 0; i < 6; i++)
            {
                game.WonPoint(points[i]);
                Assert.AreEqual(expectedScores[i], game.GetScore());
            }
        }
    }

}

File: vbnet/Tennis.Tests/TennisTests.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports Xunit

Public Class TestDataGenerator
    Implements IEnumerable(Of Object())

    Private ReadOnly _data As List(Of Object()) = New List(Of Object()) From {
        New Object() {0, 0, "Love-All"},
        New Object() {1, 1, "Fifteen-All"},
        New Object() {2, 2, "Thirty-All"},
        New Object() {3, 3, "Deuce"},
        New Object() {4, 4, "Deuce"},
        New Object() {1, 0, "Fifteen-Love"},
        New Object() {0, 1, "Love-Fifteen"},
        New Object() {2, 0, "Thirty-Love"},
        New Object() {0, 2, "Love-Thirty"},
        New Object() {3, 0, "Forty-Love"},
        New Object() {0, 3, "Love-Forty"},
        New Object() {4, 0, "Win for player1"},
        New Object() {0, 4, "Win for player2"},
        New Object() {2, 1, "Thirty-Fifteen"},
        New Object() {1, 2, "Fifteen-Thirty"},
        New Object() {3, 1, "Forty-Fifteen"},
        New Object() {1, 3, "Fifteen-Forty"},
        New Object() {4, 1, "Win for player1"},
        New Object() {1, 4, "Win for player2"},
        New Object() {3, 2, "Forty-Thirty"},
        New Object() {2, 3, "Thirty-Forty"},
        New Object() {4, 2, "Win for player1"},
        New Object() {2, 4, "Win for player2"},
        New Object() {4, 3, "Advantage player1"},
        New Object() {3, 4, "Advantage player2"},
        New Object() {5, 4, "Advantage player1"},
        New Object() {4, 5, "Advantage player2"},
        New Object() {15, 14, "Advantage player1"},
        New Object() {14, 15, "Advantage player2"},
        New Object() {6, 4, "Win for player1"},
        New Object() {4, 6, "Win for player2"},
        New Object() {16, 14, "Win for player1"},
        New Object() {14, 16, "Win for player2"}
        }

    Public Function IEnumerableOfT_GetEnumerator() As IEnumerator(Of Object()) Implements IEnumerable(Of Object()).GetEnumerator
        Return _data.GetEnumerator()
    End Function

    Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
        Return IEnumerableOfT_GetEnumerator()
    End Function
    
End Class

Public Class TennisTests
    <Theory>
    <ClassData(GetType(TestDataGenerator))>
    Public Sub Tennis1Test(ByVal p1 As Integer, ByVal p2 As Integer, ByVal expected As String)
        Dim game = New TennisGame1("player1", "player2")
        CheckAllScores(game, p1, p2, expected)
    End Sub

    <Theory>
    <ClassData(GetType(TestDataGenerator))>
    Public Sub Tennis2Test(ByVal p1 As Integer, ByVal p2 As Integer, ByVal expected As String)
        Dim game = New TennisGame2("player1", "player2")
        CheckAllScores(game, p1, p2, expected)
    End Sub

    <Theory>
    <ClassData(GetType(TestDataGenerator))>
    Public Sub Tennis3Test(ByVal p1 As Integer, ByVal p2 As Integer, ByVal expected As String)
        Dim game = New TennisGame3("player1", "player2")
        CheckAllScores(game, p1, p2, expected)
    End Sub

    Private Sub CheckAllScores(ByVal game As ITennisGame, ByVal player1Score As Integer, ByVal player2Score As Integer, ByVal expectedScore As String)
        Dim highestScore = Math.Max(player1Score, player2Score)

        For i = 0 To highestScore - 1
            If i < player1Score Then game.WonPoint("player1")
            If i < player2Score Then game.WonPoint("player2")
        Next

        Assert.Equal(expectedScore, game.GetScore())
    End Sub
End Class